### PR TITLE
k path for the original cell without standardization

### PR DESCRIPTION
**Pull request by jaemolihm**
As I discussed with @giovannipizzi during the Wannier 2022 developers meeting, I added two functions `get_path_orig_cell` and `get_explicit_k_path_orig_cell`, which does the same thing with the existing functions except that they keep the original cell: the original cell is not standardized or symmetrized.

The idea for the implementation is that linear combination of the lattice vectors does not change the 1st BZ, so only the real space rotations needs to be taken into account. I first implemented this idea in `Brillouin.jl`, a Julia implementation of the seekpath algorithm. For further information, see discussions in https://github.com/thchr/Brillouin.jl/pull/15, https://github.com/thchr/Brillouin.jl/pull/16.

If the input cell is a non-standard primitive unit cell, the returned k path is equivalent to the k path for the standard cell.
If the input cell is a supercell of a smaller primitive cell, the returned k path is that of the associated primitive cell, in the basis of supercell reciprocal lattice.

### Example: silicon
* Non-standard unit cell is rotated by 17 degrees and the basis vectors are linearly combined.
* The supercell is a 1\*1\*2 supercell of the non-standard primitive cell.
* Standard and non-standard unit cell show identical band structures.
* Supercell band structure includes an identical copy of the primitive cell band structure and additional folded bands.
* Input files available in https://github.com/jaemolihm/seekpath_orig_cell_example
image

### Tests
* `TestPaths3D_HPKOT_Orig_Cell.base_test`: Test the k path with and without standardization are equivalent up to rotation.
* `TestPaths3D_HPKOT_Orig_Cell.test_no_symmetrization`: Check that `get_path_orig_cell` does not symmetrize the input structure so that for a slightly distorted system the k points are not exactly on the high-symmetry points
* `TestExplicitPaths_Orig_Cell.test_keys`: Test output keys of `get_explicit_k_path_orig_cell`
* `TestExplicitPaths_Orig_Cell.test_path`: Test the explicit k point list for with and without standardization are equivalent up to rotation.

### CODE REVIEW

**Comment by jaemolihm** [comment]
This PR does not "Provide Transformation Matrix as output of the code" which is the title of issue #16, but it solves the related issue of calculating the k path for non-standard cells (#77, https://github.com/aiidateam/aiida-core/issues/4391).

**Comment by tkotani** [comment]
This is what I implemented several years ago.
As you noticed, we have to find out unimodular matrix which connects two sets of primitive vectors.
FYI: Related to "lattice encryption", we don't have very efficient way to find out the unimodular matrix.
Some trial and error procedure needed, I think. (this is what I had implemented).
Math of integer matrix is not so easy to handle.

**Comment by jaemolihm** [comment]
> This is what I implemented several years ago. As you noticed, we have to find out unimodular matrix which connects two sets of primitive vectors. FYI: Related to "lattice encryption", we don't have very efficient way to find out the unimodular matrix. Some trial and error procedure needed, I think. (this is what I had implemented). Math of integer matrix is not so easy to handle.

Dear @tkotani , thanks for the comments. To be clear, I did not calculate or use the integer-valued unimodular matrix. Instead, I used the orthogonal rotation matrix that describes the real-space rotation (`std_rotation_matrix` of spglib).

Here, I did not need the integer-valued matrix. If in the future one wants it for some reason, one can use L_std = R * L_inp * N (L_std: standard primitive lattice, L_inp: input lattice, R: orthogonal rotation matrix, N: integer-valued unimodular matrix). Since L_std, R, L_inp are known, N can be easily computed.

**Comment by tkotani** [comment]
Dear @jaemolihm, Thank you for your reply. I see.
Generally speaking, the same lattice is defined by different sets of primitive vectors.
The sets are related by unimodular matrix. 
 
See https://cseweb.ucsd.edu/classes/wi12/cse206A-a/lec1.pdf
Theorem 4. (page 2.) is what I said. We have to find out the matrix U.

File: docs/source/maindoc.rst [hunk 1]
```diff
@@ -61,24 +61,42 @@ The output of the function is a dictionary containing, among other quantities, t
...
+     seekpath.get_path_orig_cell
+
+If the input cell is a non-standard primitive unit cell, the returned k path is equivalent to the k path for the standard cell.
+For example, the band structure calculated along the k path for the standard and non-standard unit cells will be the same up to numerical errors.
+
+If the input cell is a supercell of a smaller primitive cell, the returned k path is that of the associated primitive cell, in the basis of supercell reciprocal lattice.
+In this case, the k points are **not** the high-symmetry points of the first Brillouin zone of the given supercell, but the high-symmetry points of the Brillouin zone of the associated primitive cell.
+
+Note that contrary to ``get_path``, ``get_path_orig_cell`` calculates the k path based on the symmetrized structure but does not symmetrize the input structure itself.
+Hence, if the symmetry of the input structure is slightly broken below the symmetry precision ``symprec``, the output k points may not be exactly on the high-symmetry k points.
```
Comment (giovannipizzi) [inline docs/source/maindoc.rst:81]:
```suggestion
**Note that contrary to ``get_path``, ``get_path_orig_cell`` calculates the k path based on the symmetrized structure but does not symmetrize the input structure itself.**
Hence, if the symmetry of the input structure is slightly broken below the symmetry precision ``symprec``, the **output k points may not be exactly on the high-symmetry k points.**
```

File: docs/source/maindoc.rst [hunk 2]
```diff
@@ -61,24 +61,42 @@ The output of the function is a dictionary containing, among other quantities, t
...
    output) and store it somewhere, together with the k-point coordinates
    and suggested band path
 
 2. You then run all your calculations using the standardized primitive cell
 
-If you already have done calculations with a non-standardized cell, you will
-then need to figure out how to remap the labeled k-points in the choice of
-cell you did.
+If you want a k-point path for a non-standardized cell, you can use the
+``get_path_orig_cell`` and ``get_explicit_k_path_orig_cell`` functions: see the above subsection.
```
Comment (giovannipizzi) [inline docs/source/maindoc.rst:99]:
```suggestion
``get_path_orig_cell`` and ``get_explicit_k_path_orig_cell`` functions: see the above subsection (and in particular, check the limitations if the symmetry of the input structure is slightly broken).
```

File: seekpath/getpaths.py [hunk 3]
```diff
@@ -325,3 +325,268 @@ def get_explicit_k_path(
...
+    recipe="hpkot",
+    threshold=1.0e-7,
+    symprec=1e-05,
+    angle_tolerance=-1.0,
+):
+    r"""
+    Return the kpoint path information for band structure given a
+    crystal structure, using the paths from the chosen recipe/reference.
+    The original unit cell is used. Standardization or symmetrization of the
+    input unit cell is not performed.
```
Comment (giovannipizzi) [inline seekpath/getpaths.py:342]:
```suggestion
    The original unit cell (i.e., the one provided in input by the user) is used. 
    Standardization or symmetrization of the input unit cell is not performed.
```

File: seekpath/getpaths.py [hunk 4]
```diff
@@ -325,3 +325,268 @@ def get_explicit_k_path(
...
+
+    If the provided unit cell is a supercell of a smaller primitive cell,
+    return the standard k path of the smaller primitive cell in the basis
+    of the supercell reciprocal lattice vectors. In this case, the k point
+    labels lose their meaning: they are not at the high-symmetry points of
+    the first BZ of the given supercell.
```
Comment (giovannipizzi) [inline seekpath/getpaths.py:348]:
```suggestion
    of the supercell reciprocal lattice vectors. Note that, in this case,
    the k-point labels are not at the high-symmetry points of
    the first BZ of the given supercell.
```

File: seekpath/getpaths.py [hunk 5]
```diff
@@ -325,3 +325,268 @@ def get_explicit_k_path(
...
+        threshold=threshold,
+        symprec=symprec,
+        angle_tolerance=angle_tolerance,
+        recipe=recipe,
+    )
+
+    is_supercell = abs(res["volume_original_wrt_prim"] - 1) > 0.1
+
+    if is_supercell:
+        print(
```
Comment (giovannipizzi) [inline seekpath/getpaths.py:425]: Replace with a warning instead, so it can be easily filtered

File: seekpath/hpkot/__init__.py [hunk 6]
```diff
@@ -120,6 +120,8 @@ def get_path(
         - ``inverse_primitive_transformation_matrix``: the inverse of the matrix :math:`P`
           (the determinant is integer and gives the ratio in volume between
           the conventional and primitive cells)
+        - ``rotation_matrix``: rotation matrix in the Cartesian space from the input
```
Comment (giovannipizzi) [inline seekpath/hpkot/__init__.py:123]:
```suggestion
        - ``rotation_matrix``: rotation matrix in Cartesian space from the input
```

File: seekpath/getpaths.py (hunk 5 above)
Comment (jaemolihm) [reply seekpath/getpaths.py]: Done

File: seekpath/getpaths.py (hunk 4 above)
Comment (jaemolihm) [reply seekpath/getpaths.py]: I did a slightly modified version: "In this case, the k-point labels lose their meaning as the corresponding k-points are not at the high-symmetry"

**Comment by jaemolihm** [comment]
I locally tested with spglib-2.0.2 and tests passed.

**Review by giovannipizzi — APPROVED** [review]
Thanks a lot, looks great!

**Comment by jaemolihm** [comment]
Thanks a lot for the review!

**Pull request merged**